File: Offer-Cpp/16.合并两个排序的链表.cpp
//
// Created by 陈建虎 on 2019-11-04.
//
//题目描述
//输入两个单调递增的链表，输出两个链表合成后的链表，当然我们需要合成后的链表满足单调不减规则。

#include <iostream>

using namespace std;

struct ListNode {
    int val;
    struct ListNode *next;

    ListNode(int x) :
            val(x), next(NULL) {
    }
};


class Solution {
public:
    ListNode *Merge(ListNode *pHead1, ListNode *pHead2) {
        ListNode *pHead = new ListNode(0);
        ListNode *pRoot = pHead;
        while (pHead1 && pHead2) {
            ListNode *newNode = new ListNode(0);
            if (pHead1->val < pHead2->val) {
                newNode->val = pHead1->val;
                pHead1 = pHead1->next;
            } else {
                newNode->val = pHead2->val;
                pHead2 = pHead2->next;
            }
            pHead->next = newNode;
            pHead = pHead->next;
        }

        while (pHead1) {
            ListNode *newNode = new ListNode(pHead1->val);
            pHead->next = newNode;
            pHead = pHead->next;
            pHead1 = pHead1->next;
        }

        while (pHead2) {
            ListNode *newNode = new ListNode(pHead2->val);
            pHead->next = newNode;
            pHead = pHead->next;
            pHead2 = pHead2->next;
        }

        return pRoot->next;
    }
};

File: offer/16.合并两个排序的链表.py
## 题目描述
# 输入两个单调递增的链表，输出两个链表合成后的链表，
# 当然我们需要合成后的链表满足单调不减规则。

# -*- coding:utf-8 -*-
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
class Solution:
    # 返回合并后列表
    def Merge(self, pHead1, pHead2):
        # write code here
        l = []
        while pHead1 and pHead2:
            if pHead1.val <= pHead2.val:
                l.append(pHead1)
                pHead1 = pHead1.next
            else:
                l.append(pHead2)
                pHead2 = pHead2.next

        while pHead1:
            l.append(pHead1)
            pHead1 = pHead1.next

        while pHead2:
            l.append(pHead2)
            pHead2 = pHead2.next
            
        if l != []:
            head = l[0]
            for i in range(len(l)-1):
                l[i].next = l[i+1]
            l[-1].next = None
            return head
            